src/python/pants/backend/go/util_rules/analyze_test_sources.go

```diff
@@ -32,23 +37,83 @@ import (
 //    test framework with a list of tests, benchmarks, examples, and fuzz
 //    targets read from source files.
 //
+// https://github.com/bazelbuild/rules_go/blob/master/go/tools/builders/generate_test_main.go
+//
 
-type TestCase struct {
+type TestFunc struct {
 	Package string `json:"package"`
 	Name    string `json:"name"`
 }
 
+type Example struct {
+	Package   string `json:"package"`
+	Name      string `json:"name"`
+	Output    string `json:"output"`
+	Unordered bool   `json:"unordered"`
+}
+
 // TestSourcesMetadata contains metadata about tests/benchmarks extracted from the parsed sources.
 // TODO: "Examples" and "fuzz targets" (Go 1.18+).
 type TestSourcesMetadata struct {
 	// Names of all functions in the test sources that heuristically look like test functions.
-	Tests []*TestCase `json:"tests,omit_empty"`
+	Tests []*TestFunc `json:"tests,omitempty"`
 
 	// Names of all functions in the test sources that heuristically look like benchmark functions.
-	Benchmarks []*TestCase `json:"benchmarks,omit_empty"`
+	Benchmarks []*TestFunc `json:"benchmarks,omitempty"`
+
+	// Testable examples. Extracted using "go/doc" package.
+	Examples []*Example `json:"examples,omitempty"`
 
 	// True if the sources already contain a `TestMain` function (which is the entry point for test binaries).
-	HasTestMain bool `json:"has_test_main"`
+	TestMain *TestFunc `json:"test_main,omitempty"`
+}
+
+// isTestFunc tells whether fn has the type of a testing function. arg
+// specifies the parameter type we look for: B, M or T.
+func isTestFunc(fn *ast.FuncDecl, arg string) bool {
+	if fn.Type.Results != nil && len(fn.Type.Results.List) > 0 ||
+		fn.Type.Params.List == nil ||
+		len(fn.Type.Params.List) != 1 ||
+		len(fn.Type.Params.List[0].Names) > 1 {
+		return false
+	}
+	ptr, ok := fn.Type.Params.List[0].Type.(*ast.StarExpr)
+	if !ok {
+		return false
+	}
+	// We can't easily check that the type is *testing.M
+	// because we don't know how testing has been imported,
+	// but at least check that it's *M or *something.M.
+	// Same applies for B and T.
+	if name, ok := ptr.X.(*ast.Ident); ok && name.Name == arg {
+		return true
+	}
+	if sel, ok := ptr.X.(*ast.SelectorExpr); ok && sel.Sel.Name == arg {
+		return true
+	}
+	return false
+}
+
+// isTest tells whether name looks like a test (or benchmark, according to prefix).
+// It is a Test (say) if there is a character after Test that is not a lower-case letter.
```

Comment:
What do you mean with `(say)`?

Reply:
That comment comes from the upstream `go` tool's source.

I edited out the final line of that comment which said: 

```
// We don't want TesticularCancer.
```

as sort of not good for a comment.

https://github.com/golang/go/blob/f0db7eae74ea235e9fbc2598252bfd46c1cc5510/src/cmd/go/internal/load/test.go#L506-L508

Reply:
Agreed on not being a good comment...maybe `It is a test if there is a character after Test that is not a lower-case letter. This avoids, for example, Testify matching.

Reply:
Done.